### ISSUE TITLE
Rule Request: [implicit-return] Support SE-0255: implicit returns in single-expression functions

### ISSUE THREAD

**Issue by BasThomas** [realm/SwiftLint#2870]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

I'm not sure if this should be an extension (and thus be a breaking change) to the currently existing `implicit-return`; [that rule](https://github.com/realm/SwiftLint/blob/406a8f20f8c55d269d52844c41382bccda868aac/Rules.md#L10067-L10073) is currently only referring to closures. This could be a separate rule (`implicit-return-single-expression-functions`?) but that could still be confusable with `implicit-return` (renaming that to `implicit-return-closures` also feels like a tough sell.

1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.

Swift introduced [implicit returns in single-expression functions](https://github.com/apple/swift-evolution/blob/master/proposals/0255-omit-return.md) in SE-0255. From there:

> Swift provides a pleasant shorthand for short closures: if a closure contains just a single expression, that expression is implicitly returned--the `return` keyword can be omitted.  We should provide this shorthand for functions as well.

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

Triggers:

```swift
var x: String { return "" }
func y() -> String { return "" }
func z() -> ((String) -> Void) { return { _ in }}
```

No triggers:

```swift
var x: String { print("hi!"); return "" }
func y() -> String {
  print("hi!")
  return "bye"
}
```

3. Should the rule be configurable, if so what parameters should be configurable?

This should be considered as an alternative to what's mentioned above (above point 1.).

Yes, it could be, assuming we can add this to the existing `implicit-return` rule:

- `single_expression_functions` (set to `false` by default)
- `closures` (set to `true` by default)

4. Should the rule be opt-in or enabled by default? Why?
   See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a rule as opt-in.

No; just like `implicit-return` is already disabled by default.

**Comment by marcelofabri** [comment]
This could also be a configuration on the existing rule

**Comment by BasThomas** [comment]
Yep, I mentioned that in 3. :)

**Comment by iliaskarim** [comment]
This could even better be a configuration on ImplicitGetterRule.

**Comment by BasThomas** [comment]
I don't see the connection between the two, as single-body expressions are also to be found in functions.

**Comment by iliaskarim** [comment]
> I don't see the connection between the two, as single-body expressions are also to be found in functions.

e.g. ImplicitGetterRule, omit_return configuration
+ ImplicitFunctionReturnRule

**Comment by Czajnikowski** [comment]
It'd be also nice to have a contrary rule - to enforce explicit returns. This may be a part of the configuration.

**Comment by danl3v** [comment]
Yes +1 to forcing explicit returns. Removing returns can make things harder to read!

**Comment by svenmuennich** [comment]
I went ahead and improved the existing rule `implicit_return` to also include functions and getters. The rule can also be configured to selectively enable it for closures, functions and/or getters.

See #2976.

**Comment by jpsim** [comment]
Pushed to master in 72e20635.

**Comment by BasThomas** [comment]
Awesome! Thanks, @svenmuennich & @jpsim :)
